File: backend/app/schemas/tutor.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.question import AnswerOption, Difficulty
from app.models.tutor import TutorSessionStatus


class TutorSessionCreateRequest(BaseModel):
    course_id: int
    week_ids: list[int] | None = None
    topic: str | None = None


class TutorTurnPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    turn_order: int
    explanation: str
    question_text: str
    option_a: str
    option_b: str
    option_c: str
    option_d: str
    difficulty_at_time: Difficulty
    document_filename: str
    page_number: int | None
    student_answer: AnswerOption | None
    is_correct: bool | None
    hint_used: bool


class TutorSessionPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    course_id: int
    topic: str | None
    difficulty: Difficulty
    correct_streak: int
    incorrect_streak: int
    questions_asked: int
    questions_correct: int
    status: TutorSessionStatus
    created_at: datetime
    turns: list[TutorTurnPublic]


class TutorAnswerRequest(BaseModel):
    selected_answer: AnswerOption


class TutorAnswerResponse(BaseModel):
    is_correct: bool
    correct_answer: AnswerOption
    previous_difficulty: Difficulty
    new_difficulty: Difficulty
    difficulty_changed: bool
    session: TutorSessionPublic


class TutorHintResponse(BaseModel):
    hint: str

File: frontend/src/types/Tutor.ts
import type { AnswerOption, Difficulty } from './Question'

export type TutorSessionStatus = 'active' | 'ended'

export interface TutorSessionCreateRequest {
  course_id: number
  week_ids?: number[] | null
  topic?: string | null
}

export interface TutorTurnPublic {
  id: number
  turn_order: number
  explanation: string
  question_text: string
  option_a: string
  option_b: string
  option_c: string
  option_d: string
  difficulty_at_time: Difficulty
  document_filename: string
  page_number: number | null
  student_answer: AnswerOption | null
  is_correct: boolean | null
  hint_used: boolean
}

export interface TutorSessionPublic {
  id: number
  course_id: number
  topic: string | null
  difficulty: Difficulty
  correct_streak: number
  incorrect_streak: number
  questions_asked: number
  questions_correct: number
  status: TutorSessionStatus
  created_at: string
  turns: TutorTurnPublic[]
}

export interface TutorAnswerRequest {
  selected_answer: AnswerOption
}

export interface TutorAnswerResponse {
  is_correct: boolean
  correct_answer: AnswerOption
  previous_difficulty: Difficulty
  new_difficulty: Difficulty
  difficulty_changed: boolean
  session: TutorSessionPublic
}

export interface TutorHintResponse {
  hint: string
}
